pretyping/pretyping.mli

```diff
@@ -130,3 +110,45 @@ val ise_pretype_gen :
 val register_constr_interp0 :
   ('r, 'g, 't) Genarg.genarg_type ->
     (unbound_ltac_var_map -> env -> evar_map -> types -> 'g -> constr * evar_map) -> unit
+
+(** Open-recursion style pretyper *)
+
+open Misctypes
+open Glob_term
+
+module ExtraEnv :
+sig
+  type t
+  val env : t -> Environ.env
+  val make_env : Environ.env -> evar_map -> t
+  val push_rel : evar_map -> rel_declaration -> t -> t
+end
+(** A wrapper around [Environ.env] to generate more efficiently named contexts
+    of fresh evars. *)
+
+type 'a pretype_fun = Loc.t option -> bool -> type_constraint -> ExtraEnv.t -> evar_map ref -> (int * ltac_var_map) -> 'a
```

Comment:
May I suggest `... -> evar_map -> evar_map * 'a`?

Also, I quite don't see why one should "hardcode" here `ltac_var_map`.

Reply:
Both are going to be changed eventually. The pretyper needs to be ported to the tactic monad, which will solve many bugs, and the Ltac specific part should be factored out using either dynamic data or a functor.